test/local-hub/external_storage_backend_test.go

```diff
@@ -0,0 +1,581 @@
+			},
+			locked:                  "",
+			lockedInExternalBackend: "",
+		},
+		{
+			alias: ptr.String("parent"),
+			backendContext: map[string]interface{}{
+				"provider": "dotenv",
+			},
+			dataInGQL: map[string]interface{}{
+				"name": "Darth Vader",
+			},
+			dataInExternalBackend: map[string]interface{}{
+				"name": "Darth Vader",
+			},
+			locked:                  "",
+			lockedInExternalBackend: "",
+		},
+	})
+}
+
+func registerExternalDotenvStorage(ctx context.Context, t *testing.T, cli *local.Client, srvAddr string) (gqllocalapi.CreateTypeInstanceOutput, func(t *testing.T)) {
+	t.Helper()
+
+	ti, err := cli.CreateTypeInstances(ctx, fixExternalDotenvStorage(srvAddr))
+	require.NoError(t, err)
+	require.Len(t, ti, 1)
+	dotenvHubStorage := ti[0]
+
+	return dotenvHubStorage, func(t *testing.T) {
+		err = cli.DeleteTypeInstance(ctx, dotenvHubStorage.ID)
+		require.NoError(t, err)
+	}
+}
+
+func fixExternalDotenvStorage(addr string) *gqllocalapi.CreateTypeInstancesInput {
+	return &gqllocalapi.CreateTypeInstancesInput{
+		TypeInstances: []*gqllocalapi.CreateTypeInstanceInput{
+			{
+				CreatedBy: ptr.String("manually"),
+				TypeRef: &gqllocalapi.TypeInstanceTypeReferenceInput{
+					Path:     "cap.type.example.filesystem.storage",
+					Revision: "0.1.0",
+				},
+				Value: storageSpec{
+					URL:         ptr.String(addr),
+					AcceptValue: ptr.Bool(true),
+					ContextSchema: ptr.String(heredoc.Doc(`
+				      {
+				      	"$id": "#/properties/contextSchema",
+				      	"type": "object",
+				      	"properties": {
+				      		"provider": {
+				      			"$id": "#/properties/contextSchema/properties/name",
+				      			"type": "string",
+								  "enum": [
+									"aws_secretsmanager",
+									"dotenv"
+								  ]
+				      		}
+				      	},
+				      	"additionalProperties": true
+				      }`)),
+				},
+			},
+		},
+		UsesRelations: []*gqllocalapi.TypeInstanceUsesRelationInput{},
+	}
+}
+
+type externalData struct {
+	Value    string
+	LockedBy *string
+}
+
+func getDataDirectlyFromStorage(t *testing.T, addr string, details []gqllocalapi.TypeInstance) map[string]externalData {
+	t.Helper()
+	url := strings.Split(addr, ":")
+	assert.Equal(t, len(url), 2)
+	conn, err := grpc.Dial(fmt.Sprintf("[%s]:%s", url[0], url[1]), grpc.WithInsecure())
+	require.NoError(t, err)
+
+	ctx := context.Background()
+	client := pb.NewStorageBackendClient(conn)
+
+	var out = map[string]externalData{}
+	for _, ti := range details {
+		val, err := client.GetValue(ctx, &pb.GetValueRequest{
+			TypeInstanceId:  ti.ID,
+			ResourceVersion: uint32(ti.LatestResourceVersion.ResourceVersion),
+		})
+		if err != nil {
+			t.Logf("error while getting value from storage for TypeInstance %s", ti.ID)
+			continue
+		}
+
+		locked, err := client.GetLockedBy(ctx, &pb.GetLockedByRequest{
+			TypeInstanceId: ti.ID,
+		})
+		if err != nil {
+			t.Logf("error while getting value of locked by from storage for TypeInstance %s", ti.ID)
+			continue
```

Comment:
instead of `t.Logf` we should use `require.NoError`

Reply:
Actually, we pass to this function results from `ListTypeInstances`, and not all of the TypeInstnaces has some data in the external storage - that is why we used here the `t.Logf` instead of `require.NoError`

Reply:
You mean that sometimes errors are expected? If this is validated somewhere else then I'm fine with that for now 👍